code/modules/xenoarchaeology/traits/xenoartifact_traits.dm

```diff
@@ -0,0 +1,123 @@
+///Xenoartifact traits, datum-ised
+/datum/xenoartifact_trait
+	///Acts as a descriptor for when examining. Also used for naming stuff in the labeler. Keep these short.
+	var/desc
+	///Used when labeler needs a name and trait is too sneaky to have a descriptor when examining.
+	var/label_name
+	///Something briefly explaining it in IG terms or a pun.
+	var/label_desc
+	///Asscoiated flags for artifact typing
+	var/flags = NONE
+	///Other traits the original trait wont work with. Referenced when generating traits.
+	var/list/blacklist_traits = list()
+	///Weight in trait list, most traits wont change this
+	var/weight = 50
+
+///Proc used to compile trait weights into a list
+/proc/compile_artifact_weights(path)
+	if(!ispath(path))
+		return
+	var/list/temp = subtypesof(path)
+	var/list/weighted = list()
+	for(var/datum/xenoartifact_trait/T as() in temp)
+		weighted += list((T) = initial(T.weight))
+	return weighted
+
+///Compile a blacklist of traits from a given flag/s
+/proc/compile_artifact_blacklist(var/flags)
+	var/list/output = list()
+	for(var/datum/xenoartifact_trait/T as() in GLOB.xenoa_all_traits)	
+		if(!(initial(T.flags) & flags))
+			output += T
+	return output
+
+//Activator signal shenanignas 
+/datum/xenoartifact_trait/activator
+	///How much an activator trait can output on a standard, modified by the artifacts charge_req and circumstances.
+	var/charge
+	///which signals trait responds to
+	var/list/signals
+	///Not used outside of signal handle, please
+	var/obj/item/xenoartifact/xenoa
+
+/datum/xenoartifact_trait/activator/proc/calculate_charge(obj/item/xenoartifact/X)
+	return
```

Comment:
What is this actually intended to do? I've seen it used in lots of different ways and a lot of them don't seem to be calculating the charge of something, some of them do processing or other weird things. Is this the activation effect?

Reply:
They call the artifact to activate after calculating charge.
Here you can see ``impact`` calculating force from charge before activating.
```
/datum/xenoartifact_trait/activator/impact/calculate_charge(datum/source, obj/item/thing, mob/user, atom/target)
	var/obj/item/xenoartifact/X = source
	charge = charge*((thing?.force || 10)*0.1)
	X.default_activate(charge, user, target)
```
It's not always 100% consistent but it should be fine IMO.